### ISSUE TITLE
Rancher 2 "Failed to connect to proxy" when importing cluster

### ISSUE THREAD

**Issue by jmreicha** [rancher/rancher#14449]
**Rancher versions:**
rancher/server or rancher/rancher: `2.0.5`
rancher/agent or rancher/rancher-agent: `2.0.5`

**Docker version: (`docker version`,`docker info` preferred)** 

```
Containers: 25
 Running: 24
 Paused: 0
 Stopped: 1
Images: 50
Server Version: 18.04.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.4.77-rockchip-ayufan-136
Operating System: Ubuntu 16.04.4 LTS
OSType: linux
Architecture: aarch64
CPUs: 4
Total Memory: 1.883GiB
Name: kube-node-3
ID: ULJM:6CKT:LC3W:II7M:F6GV:MZDP:5JI2:UCAA:NQSV:YNML:77IV:VFBG
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Operating system and kernel: (`cat /etc/os-release`, `uname -r` preferred)**

```
4.4.77-rockchip-ayufan-136
```

**Type/provider of hosts: (VirtualBox/Bare-metal/AWS/GCE/DO)** `Bare metal`

**Setup details: (single node rancher vs. HA rancher, internal DB vs. external DB)** `Single node container`

**Steps to Reproduce:**

Spin up a fresh Rancher server container

```
docker run -d --restart=unless-stopped \
  -p 80:80 -p 443:443 --name rancher-server \
  -v /var/lib/rancher:/var/lib/rancher \
  rancher/rancher:v2.0.5 \
  --acme-domain rancher.example.com
```

Then import a Kubernetes (1.11) cluster.

**Results:**

```
time="2018-07-08T15:47:02Z" level=info msg="Rancher agent version v2.0.5-rc2 is starting"
time="2018-07-08T15:47:02Z" level=info msg="Option customConfig=map[label:map[] address:192.168.1.207 internalAddress: roles:[]]"
time="2018-07-08T15:47:02Z" level=info msg="Option etcd=false"
time="2018-07-08T15:47:02Z" level=info msg="Option controlPlane=false"
time="2018-07-08T15:47:02Z" level=info msg="Option worker=false"
time="2018-07-08T15:47:02Z" level=info msg="Option requestedHostname=kube-node-1"
time="2018-07-08T15:47:02Z" level=info msg="Connecting to wss://rancher.example.com/v3/connect with token f5qfq6vj4vbmdgjszjpc8vbg4wn84gphsdvmkscrfcf8xkns2hbrzl"
time="2018-07-08T15:47:02Z" level=info msg="Connecting to proxy" url="wss://rancher.example.com/v3/connect"
time="2018-07-08T15:47:02Z" level=error msg="Failed to connect to proxy" error="websocket: bad handshake"
time="2018-07-08T15:47:02Z" level=error msg="Failed to connect to proxy" error="websocket: bad handshake"
time="2018-07-08T15:47:12Z" level=info msg="Connecting to wss://rancher.example.com/v3/connect with token f5qfq6vj4vbmdgjszjpc8vbg4wn84gphsdvmkscrfcf8xkns2hbrzl"
time="2018-07-08T15:47:12Z" level=info msg="Connecting to proxy" url="wss://rancher.example.com/v3/connect"
time="2018-07-08T15:47:12Z" level=error msg="Failed to connect to proxy" error="websocket: bad handshake"
time="2018-07-08T15:47:12Z" level=error msg="Failed to connect to proxy" error="websocket: bad handshake"
```

I am able to curl the URL from inside the cattle-node-agent container so I'm not really sure where it is getting stuck.

**Comment by superseb** [comment]
What cluster are you importing? Is `rancher.example.com` resolving directly to the host running the `rancher/rancher` or is there something in between? Did you run `curl https://rancher.example.com`?

**Comment by jmreicha** [comment]
`rancher.example.com` is resolving, nothing in-between.  Curl works.  I think this might be due to some settings I was attempting to change after creating the cluster import.

I got things working so I think it has something to do with my non standard configuration, so I'll close this one.

**Comment by steve-todorov** [comment]
For anybody else who might experience this issue - this could be a firewall issue (at least that's what turned out to be with our configuration)

**Comment by TheBoroer** [comment]
@steve-todorov firewall issue on rancher server? or something with the cluster nodes you were importing?

**Comment by steve-todorov** [comment]
In our case we were adding a new node, but it was receiving the same error. It turned out to be an issue related to the `FORWARD` chain and the `flannel.1` interface (if I recall correctly) which was causing some weird behavior.

**Comment by bluemalkin** [comment]
I'm running into this issue as well on GCP and I'm out of ideas how to solve it. Firewall is open on 80/443... Any tips appreciated.

**Comment by dsuarezf** [comment]
I have the same problem too:

level=error msg="Failed to connect to proxy" error="x509: certificate signed by unknown authority"

**Comment by sosoriov** [comment]
Hi @steve-todorov 

I'm having similar issues with a cluster that has been upgraded to Rancher 2.2.6. Can you point me out or elaborate a bit more about what you found Calico/Flannel related?

Thanks

**Comment by steve-todorov** [comment]
@sosoriov In our case, we had a bit more restrictive firewall rules on our nodes which were causing the issue. The way we found it was by stopping the firewall and making all chains "ACCEPT" everything. At that point everything worked just fine. 

We did some firewall debugging and noticed the firewall was configured to start before docker (which is how we needed it to be) and the therefore flannel interface(s) were going up late into the chain. That somehow was messing up the firewall rules causing strange connection failures.

The way we fixed it is by switching all rules using the `flannel.1` interface to use the flannel IP subnet instead. Afterwards everything worked as expected.

I would suggest you try to entirely disable the firewall on the machine (and any connected cluster nodes), reboot (all machines) and see if it magically works. If it does - the issues is somewhere in your firewall rules. If it doesn't - maybe you have another issue?
